examples/pytorch/adaptive_sampling/adaptive_sampling.py

```diff
@@ -0,0 +1,475 @@
+            rel_graphs=rel_graphs,
+            seed_map=seed_map)
+
+    def stepback(self, curr_frontier, *auxiliary):
+        """Function that takes in the node set in the current layer, and returns the
+        neighbors of each node.
+
+        Parameters
+        ----------
+        curr_frontier : Tensor
+        auxiliary : any auxiliary data yielded by the sampler
+
+        Returns
+        -------
+        neighbor_nodes, incoming_edges, num_neighbors, auxiliary: Tensor, Tensor, Tensor, any
+            num_neighbors[i] contains the number of neighbors generated for curr_frontier[i]
+
+            neighbor_nodes[sum(num_neighbors[0:i]):sum(num_neighbors[0:i+1])] contains the actual
+            neighbors as node IDs in the original graph for curr_frontier[i].
+
+            incoming_edges[sum(num_neighbors[0:i]):sum(num_neighbors[0:i+1])] contains the actual
+            incoming edges as edge IDs in the original graph for curr_frontier[i], or -1 if the
+            edge does not exist, or if we don't care about the edge, in the original graph.
+
+            auxiliary could be of any type containing block-specific additional data.
+        """
+        raise NotImplementedError
+
+
+class AdaptGenerator(IterativeGenerator):
+    """
+    Nodeflow generator used for adaptive sampling
+
+    """
+
+    def __init__(self, graph, num_blocks, node_feature=None, sampler=None, num_workers=0, coalesce=False,
+                 sampler_weights=None, layer_nodes=None):
+        self.node_feature = node_feature.double()
+        adj = graph.adjacency_matrix_scipy()
+        adj.data = np.ones(adj.nnz)
+        self.norm_adj = normalize_adj(adj).tocsr()
+        self.layer_nodes = layer_nodes
+        if sampler_weights is not None:
+            self.sample_weights = sampler_weights
+        else:
+            self.sample_weights = torch.randn((input_size, 2), dtype=torch.float64, requires_grad=True)
+            nn.init.xavier_uniform_(self.sample_weights)
+        super(AdaptGenerator, self).__init__(graph, num_blocks, sampler, num_workers, coalesce)
+
+    def stepback(self, curr_frontier, layer_index, *auxiliary):
+        # Relies on that the same dst node of in_edges are contiguous, and the dst nodes
+        # are ordered the same as curr_frontier.
+
+        is_sparse = False
+        sample_weights = self.sample_weights
+        layer_size = self.layer_nodes[layer_index]
+        current_layer_feature = self.node_feature[curr_frontier]
+
+        src, des, eid = self.graph.in_edges(curr_frontier, form='all')
+        neighbor_nodes = torch.unique(torch.cat((curr_frontier, src), dim=0), sorted=False)
+        sparse_adj = self.norm_adj[curr_frontier, :][:, neighbor_nodes]
+        square_adj = sparse_adj.multiply(sparse_adj).sum(0)
+        if is_sparse:
+            sparse_adj = sparse_adj.tocoo()
+            tensor_adj = torch.sparse.DoubleTensor(sparse_adj.row, sparse_adj.col, sparse_adj.data)
```

Comment:
Use `float`?  I noticed that your code mostly uses `double`.  Currently DGL only support `float32` kernels so you need to switch to `float32` in order to use builtin kernels (to speed up; related to `recv_func`).